Allow video scrapers/NFO files/library import to add any artwork type…

rmrector opened the pull request:
… for items

<!--- Provide a general summary of your change in the Title above -->

## Description
Scrapers (along with friends NFO files and library importing from single file XML export) can set any type of artwork (clearlogo, landscape, etc) for items in the video library.

This is such a simple loop change that doesn't disturb the beast that is scrapers, I so very much wish I had found it a couple of years ago when I first went looking for it.

## Motivation and Context
The video library has long supported freely named artwork for items in most other places, now scrapers can set this artwork from the beginning.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->
By hand, I modified the new demo Python scrapers and a classic scraper to set some extra artwork, and modified a couple of NFO files to match, then ensuring the results are what I expect. I also refreshed a few items with a regular scraper and ensured they still set their expected artwork.

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] Improvement (non-breaking change which improves existing functionality)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

rmrector committed:
Allow video scrapers/NFO files/library import to add any artwork type for items

rmrector commented:
The build failure for Win-64 looks like it was caused by something else.

hudokkow commented:
Sorry, this is not my territory.
jenkins failure indeed seems unrelated. How the hell do I remove my name from review request list?

rmrector commented:
Anyone have any thoughts on this? This is kind of important, a lot of folks like extended artwork and this has been a big hole in artwork support for awhile now (relying on an add-on seems unnecessary, scrapers can add some artwork so they should be able to add all of it).

One hole is left, Kodi doesn't import extended artwork from image files next to media. I've got Python code that identifies these images just fine, but for some reason I just cannot translate it to Kodi core (C++ noob, gah!).

zag2me commented:
@DaveTBlake was looking at imports and exports recently

DaveTBlake commented:
For music, scraping of video is different, but I am interested in how we can handle art better in core rather than leave it to addons. There is also https://github.com/xbmc/xbmc/pull/11755, which could be of interest.

rmrector commented:
@phate89 You've been hard at work on bigger parts of the scrapers, what do you think of this? This code sits up a level higher, as a filter on the data from the scraper before it makes it into the database, and is still useful to your open PR #12661.

@MartijnKaijser With all your work on Artwork Downloader, what do you think of scrapers being able to add more artwork?

phate89 commented on xbmc/video/VideoInfoScanner.cpp:
```diff
@@ -1397,6 +1378,18 @@ namespace VIDEO
 
+    // add online art
+    std::vector<CScraperUrl::SUrlEntry>& scraperArt = pItem->GetVideoInfoTag()->m_strPictureURL.m_url;
+    for (auto i = scraperArt.begin(); i != scraperArt.end(); ++i)
```
You could use a range-based for like this `for (const auto& a : pItem->GetVideoInfoTag()->m_strPictureURL.m_url)`

phate89 commented:
I'm not an expert of artwork code but it seems fine and also simpler. Why did you move the loop after fanart gathering?

rmrector commented:
I moved the loop because it can also provide a "fanart" image, but the original fanart section still offers an extra feature, the rarely-used fanart colors for skins, so they should take priority.

I just added a bit that covers what the original `art.empty()` check did for XML scrapers: it considers artwork with an empty aspect to be "thumb", which is a reasonable assumption and matches scraper behavior for episodes, or "poster" for music videos, which is a quick hack to support the XML music video scrapers that share scraper bits with music album scrapers.

zag2me commented:
Does this work for music scrapers also?

phate89 commented:
@rmrector I really don't know anything but I always tought that fanart was only in `m_fanart`..
We explicitely drop it from `arTypes` here https://github.com/rmrector/xbmc/blob/f489c22bd92aa374c17b0de62ccaef946c57236f/xbmc/video/VideoInfoScanner.cpp#L1338
But I think it make sense to be able to provide all artwork from one place.. maybe @xhaggi know better
@zag2me these are only video changes. No idea if the same could be done in music part

rmrector commented:
@zag2me The music library doesn't support the same wide selection of artwork as the video library in the first place, so it can't be extended this simply and is outside the scope of this PR.

@phate89 This new code doesn't use artTypes when processing scraper artwork (artTypes is what held back the original code), but accepts the first image of each art type/"aspect" the scraper provides, so if the scraper sets an image aspect to "fanart" in `addAvailableThumb` then it will show up here. I agree that all artwork should work the same, so I didn't want to add an exception to ignore "fanart" here, and I don't want to disturb XML scrapers, so I kept the existing m_fanart as well. If it weren't for that rarely-used feature of fanart, the colors (I like the idea, but I think they should be attached to the TV show and occasionally season rather than each fanart), I'd say Python scrapers shouldn't even have a `setAvailableFanart`. The only web service that has that info is TheTVDB, and their new v2 API (v1 will disappear soon) doesn't seem to provide the color anymore (nor their beta website), so maybe it should even go away anyway?

I've also push another change to handle season artwork, which is in another section of code. One thing about season artwork is that it doesn't work from Python scrapers in master (with notspiff's initial Python scraper), but does look to be addressed in your PR. I haven't given it a test yet, but I intend to soon; the scraper process has some goofiness with artwork, and I think the Python scraper interface can smooth some of it out.

notspiff commented on xbmc/video/VideoInfoScanner.cpp:
```diff
@@ -1463,24 +1460,17 @@ namespace VIDEO
     {
-      thumb = CScraperUrl::GetThumbURL(pItem->GetVideoInfoTag()->m_strPictureURL.GetFirstThumb(type));
-      if (!thumb.empty())
+      if (thumb.find("http://") == std::string::npos &&
```
please update for the new world of https everywhere. and use StringUtils::StartsWith

rmrector replied:
The find in the line below catches https:// and every other string that has a protocol, this line is unnecessary altogether. I can remove this line and combine the two nested ifs, but I don't see a "Find"-like method in the StringUtils. Is there one somewhere?

notspiff replied:
StringUtils::Contains(..) but really using find for stuff anywhere in string is fine by me. It was the anchoring to front that was most important in my request and startswith conveys that a bit quicker to the reader of code.

notspiff reviewed (changes requested):
fix the very minors and i will merge this. sorry for the long wait!

rmrector commented:
@notspiff ping.

notspiff commented:
jenkins build this please

notspiff commented:
@xhaggi i will do a follow up with the cosmetics, let's get this in now.

olympia commented:
Cool!
Would someone explain to me (who is unable to read the code) what to change / how to use this in the scrapers?
